// @vtex/diagnostics-nodejs/dist/types/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trace = exports.context = exports.propagation = exports.SpanKind = exports.SpanStatusCode = exports.Resource = exports.LoggerProvider = exports.MetricReader = exports.MeterProvider = exports.SamplingDecision = void 0;
var sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
Object.defineProperty(exports, "SamplingDecision", { enumerable: true, get: function () { return sdk_trace_base_1.SamplingDecision; } });
var sdk_metrics_1 = require("@opentelemetry/sdk-metrics");
Object.defineProperty(exports, "MeterProvider", { enumerable: true, get: function () { return sdk_metrics_1.MeterProvider; } });
Object.defineProperty(exports, "MetricReader", { enumerable: true, get: function () { return sdk_metrics_1.MetricReader; } });
var sdk_logs_1 = require("@opentelemetry/sdk-logs");
Object.defineProperty(exports, "LoggerProvider", { enumerable: true, get: function () { return sdk_logs_1.LoggerProvider; } });
var resources_1 = require("@opentelemetry/resources");
Object.defineProperty(exports, "Resource", { enumerable: true, get: function () { return resources_1.Resource; } });
var api_1 = require("@opentelemetry/api");
Object.defineProperty(exports, "SpanStatusCode", { enumerable: true, get: function () { return api_1.SpanStatusCode; } });
Object.defineProperty(exports, "SpanKind", { enumerable: true, get: function () { return api_1.SpanKind; } });
Object.defineProperty(exports, "propagation", { enumerable: true, get: function () { return api_1.propagation; } });
Object.defineProperty(exports, "context", { enumerable: true, get: function () { return api_1.context; } });
Object.defineProperty(exports, "trace", { enumerable: true, get: function () { return api_1.trace; } });
