// @vtex/diagnostics-semconv/dist/cjs/attributes.cjs
'use strict';

//----------------------------------------------------------------------------------------------------------
// Code generated from semantic convention specification. DO NOT EDIT.
//----------------------------------------------------------------------------------------------------------
/**
 * VTEX account name
 *
 * @note VTEX standard account name.
 */
// Type: string
// Stability: stable
const ATTR_VTEX_ACCOUNT_NAME = 'vtex.account.name';
const ATTR_VTEX_ACCOUNT_NAME_DESCRIPTION = 'VTEX account name';
/**
 * VTEX account name
 *
 * @note The parent is the main account of one client.
 */
// Type: string
// Stability: stable
const ATTR_VTEX_ACCOUNT_PARENT_NAME = 'vtex.account.parent.name';
const ATTR_VTEX_ACCOUNT_PARENT_NAME_DESCRIPTION = 'VTEX account name';
/**
 * VTEX Diagnostics lib version
 *
 * @note Required
 */
// Type: string
// Stability: stable
const ATTR_VTEX_DIAGNOSTICS_VERSION = 'vtex.diagnostics.version';
const ATTR_VTEX_DIAGNOSTICS_VERSION_DESCRIPTION = 'VTEX Diagnostics lib version';
/**
 * VTEX Identity user id
 *
 * @note VTEX identity system user id
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IDENTITY_USER_ID = 'vtex.identity.user.id';
const ATTR_VTEX_IDENTITY_USER_ID_DESCRIPTION = 'VTEX Identity user id';
/**
 * VTEX Identity user name
 *
 * @note Vtex identity user name
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IDENTITY_USER_NAME = 'vtex.identity.user.name';
const ATTR_VTEX_IDENTITY_USER_NAME_DESCRIPTION = 'VTEX Identity user name';
/**
 * VTEX guid operation
 *
 * @note Each request is one VTEX operation
 */
// Type: string
// Stability: stable
const ATTR_VTEX_OPERATION_ID = 'vtex.operation.id';
const ATTR_VTEX_OPERATION_ID_DESCRIPTION = 'VTEX guid operation';
/**
 * VTEX user agent
 *
 * @note Vtex default user agent convention
 */
// Type: string
// Stability: stable
const ATTR_VTEX_USER_AGENT = 'vtex.user_agent';
const ATTR_VTEX_USER_AGENT_DESCRIPTION = 'VTEX user agent';
/**
 * VTEX IO app author type, first-party or third-party apps
 *
 * @note Indicates what kind of group the vendor of this regular app falls into–if it's `vtex`, the author type will be `1p`, if it's something else, the author type will be `3p`.
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_APP_AUTHOR_TYPE = 'vtex_io.app.author-type';
const ATTR_VTEX_IO_APP_AUTHOR_TYPE_DESCRIPTION = 'VTEX IO app author type, first-party or third-party apps';
/**
 * VTEX IO app ID
 *
 * @note String in the format `[vendor].[name]@[version]`
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_APP_ID = 'vtex_io.app.id';
const ATTR_VTEX_IO_APP_ID_DESCRIPTION = 'VTEX IO app ID';
/**
 * VTEX IO app name
 *
 * @note Human-readable identifier for this app
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_APP_NAME = 'vtex_io.app.name';
const ATTR_VTEX_IO_APP_NAME_DESCRIPTION = 'VTEX IO app name';
/**
 * VTEX IO app regular or infra apps
 *
 * @note Indicates what kind of workload this is in the eyes of the VTEX IO infrastructure: regular apps (the ones created/maintained by VTEX IO users) or infrastructure (the ones created/maintained by the App Platform team).
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_APP_TYPE = 'vtex_io.app.type';
const ATTR_VTEX_IO_APP_TYPE_DESCRIPTION = 'VTEX IO app regular or infra apps';
/**
 * VTEX IO app vendor
 *
 * @note The tenant where this app was published
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_APP_VENDOR = 'vtex_io.app.vendor';
const ATTR_VTEX_IO_APP_VENDOR_DESCRIPTION = 'VTEX IO app vendor';
/**
 * VTEX IO app version
 *
 * @note App version in the semver format `[major].[minor].[patch]`
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_APP_VERSION = 'vtex_io.app.version';
const ATTR_VTEX_IO_APP_VERSION_DESCRIPTION = 'VTEX IO app version';
/**
 * VTEX IO account workspace name
 *
 * @note Alphanumerical string beginning with lowercase letter and may contain `-` and/or `_`. Identifies a VTEX IO variant of an account/tenant.
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_WORKSPACE_NAME = 'vtex_io.workspace.name';
const ATTR_VTEX_IO_WORKSPACE_NAME_DESCRIPTION = 'VTEX IO account workspace name';
/**
 * VTEX IO Account workspace type dev or prod
 *
 * @note Identifies whether the workspace is a production one (it may receive storefront traffic and can be promoted) or a development one (it may be used for development/testing purposes and cannot be promoted).
 */
// Type: string
// Stability: stable
const ATTR_VTEX_IO_WORKSPACE_TYPE = 'vtex_io.workspace.type';
const ATTR_VTEX_IO_WORKSPACE_TYPE_DESCRIPTION = 'VTEX IO Account workspace type dev or prod';

exports.ATTR_VTEX_ACCOUNT_NAME = ATTR_VTEX_ACCOUNT_NAME;
exports.ATTR_VTEX_ACCOUNT_NAME_DESCRIPTION = ATTR_VTEX_ACCOUNT_NAME_DESCRIPTION;
exports.ATTR_VTEX_ACCOUNT_PARENT_NAME = ATTR_VTEX_ACCOUNT_PARENT_NAME;
exports.ATTR_VTEX_ACCOUNT_PARENT_NAME_DESCRIPTION = ATTR_VTEX_ACCOUNT_PARENT_NAME_DESCRIPTION;
exports.ATTR_VTEX_DIAGNOSTICS_VERSION = ATTR_VTEX_DIAGNOSTICS_VERSION;
exports.ATTR_VTEX_DIAGNOSTICS_VERSION_DESCRIPTION = ATTR_VTEX_DIAGNOSTICS_VERSION_DESCRIPTION;
exports.ATTR_VTEX_IDENTITY_USER_ID = ATTR_VTEX_IDENTITY_USER_ID;
exports.ATTR_VTEX_IDENTITY_USER_ID_DESCRIPTION = ATTR_VTEX_IDENTITY_USER_ID_DESCRIPTION;
exports.ATTR_VTEX_IDENTITY_USER_NAME = ATTR_VTEX_IDENTITY_USER_NAME;
exports.ATTR_VTEX_IDENTITY_USER_NAME_DESCRIPTION = ATTR_VTEX_IDENTITY_USER_NAME_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_AUTHOR_TYPE = ATTR_VTEX_IO_APP_AUTHOR_TYPE;
exports.ATTR_VTEX_IO_APP_AUTHOR_TYPE_DESCRIPTION = ATTR_VTEX_IO_APP_AUTHOR_TYPE_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_ID = ATTR_VTEX_IO_APP_ID;
exports.ATTR_VTEX_IO_APP_ID_DESCRIPTION = ATTR_VTEX_IO_APP_ID_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_NAME = ATTR_VTEX_IO_APP_NAME;
exports.ATTR_VTEX_IO_APP_NAME_DESCRIPTION = ATTR_VTEX_IO_APP_NAME_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_TYPE = ATTR_VTEX_IO_APP_TYPE;
exports.ATTR_VTEX_IO_APP_TYPE_DESCRIPTION = ATTR_VTEX_IO_APP_TYPE_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_VENDOR = ATTR_VTEX_IO_APP_VENDOR;
exports.ATTR_VTEX_IO_APP_VENDOR_DESCRIPTION = ATTR_VTEX_IO_APP_VENDOR_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_VERSION = ATTR_VTEX_IO_APP_VERSION;
exports.ATTR_VTEX_IO_APP_VERSION_DESCRIPTION = ATTR_VTEX_IO_APP_VERSION_DESCRIPTION;
exports.ATTR_VTEX_IO_WORKSPACE_NAME = ATTR_VTEX_IO_WORKSPACE_NAME;
exports.ATTR_VTEX_IO_WORKSPACE_NAME_DESCRIPTION = ATTR_VTEX_IO_WORKSPACE_NAME_DESCRIPTION;
exports.ATTR_VTEX_IO_WORKSPACE_TYPE = ATTR_VTEX_IO_WORKSPACE_TYPE;
exports.ATTR_VTEX_IO_WORKSPACE_TYPE_DESCRIPTION = ATTR_VTEX_IO_WORKSPACE_TYPE_DESCRIPTION;
exports.ATTR_VTEX_OPERATION_ID = ATTR_VTEX_OPERATION_ID;
exports.ATTR_VTEX_OPERATION_ID_DESCRIPTION = ATTR_VTEX_OPERATION_ID_DESCRIPTION;
exports.ATTR_VTEX_USER_AGENT = ATTR_VTEX_USER_AGENT;
exports.ATTR_VTEX_USER_AGENT_DESCRIPTION = ATTR_VTEX_USER_AGENT_DESCRIPTION;
