// @vtex/diagnostics-semconv/dist/cjs/index.cjs
'use strict';

var attributes = require('./attributes.cjs');



exports.ATTR_VTEX_ACCOUNT_NAME = attributes.ATTR_VTEX_ACCOUNT_NAME;
exports.ATTR_VTEX_ACCOUNT_NAME_DESCRIPTION = attributes.ATTR_VTEX_ACCOUNT_NAME_DESCRIPTION;
exports.ATTR_VTEX_ACCOUNT_PARENT_NAME = attributes.ATTR_VTEX_ACCOUNT_PARENT_NAME;
exports.ATTR_VTEX_ACCOUNT_PARENT_NAME_DESCRIPTION = attributes.ATTR_VTEX_ACCOUNT_PARENT_NAME_DESCRIPTION;
exports.ATTR_VTEX_DIAGNOSTICS_VERSION = attributes.ATTR_VTEX_DIAGNOSTICS_VERSION;
exports.ATTR_VTEX_DIAGNOSTICS_VERSION_DESCRIPTION = attributes.ATTR_VTEX_DIAGNOSTICS_VERSION_DESCRIPTION;
exports.ATTR_VTEX_IDENTITY_USER_ID = attributes.ATTR_VTEX_IDENTITY_USER_ID;
exports.ATTR_VTEX_IDENTITY_USER_ID_DESCRIPTION = attributes.ATTR_VTEX_IDENTITY_USER_ID_DESCRIPTION;
exports.ATTR_VTEX_IDENTITY_USER_NAME = attributes.ATTR_VTEX_IDENTITY_USER_NAME;
exports.ATTR_VTEX_IDENTITY_USER_NAME_DESCRIPTION = attributes.ATTR_VTEX_IDENTITY_USER_NAME_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_AUTHOR_TYPE = attributes.ATTR_VTEX_IO_APP_AUTHOR_TYPE;
exports.ATTR_VTEX_IO_APP_AUTHOR_TYPE_DESCRIPTION = attributes.ATTR_VTEX_IO_APP_AUTHOR_TYPE_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_ID = attributes.ATTR_VTEX_IO_APP_ID;
exports.ATTR_VTEX_IO_APP_ID_DESCRIPTION = attributes.ATTR_VTEX_IO_APP_ID_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_NAME = attributes.ATTR_VTEX_IO_APP_NAME;
exports.ATTR_VTEX_IO_APP_NAME_DESCRIPTION = attributes.ATTR_VTEX_IO_APP_NAME_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_TYPE = attributes.ATTR_VTEX_IO_APP_TYPE;
exports.ATTR_VTEX_IO_APP_TYPE_DESCRIPTION = attributes.ATTR_VTEX_IO_APP_TYPE_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_VENDOR = attributes.ATTR_VTEX_IO_APP_VENDOR;
exports.ATTR_VTEX_IO_APP_VENDOR_DESCRIPTION = attributes.ATTR_VTEX_IO_APP_VENDOR_DESCRIPTION;
exports.ATTR_VTEX_IO_APP_VERSION = attributes.ATTR_VTEX_IO_APP_VERSION;
exports.ATTR_VTEX_IO_APP_VERSION_DESCRIPTION = attributes.ATTR_VTEX_IO_APP_VERSION_DESCRIPTION;
exports.ATTR_VTEX_IO_WORKSPACE_NAME = attributes.ATTR_VTEX_IO_WORKSPACE_NAME;
exports.ATTR_VTEX_IO_WORKSPACE_NAME_DESCRIPTION = attributes.ATTR_VTEX_IO_WORKSPACE_NAME_DESCRIPTION;
exports.ATTR_VTEX_IO_WORKSPACE_TYPE = attributes.ATTR_VTEX_IO_WORKSPACE_TYPE;
exports.ATTR_VTEX_IO_WORKSPACE_TYPE_DESCRIPTION = attributes.ATTR_VTEX_IO_WORKSPACE_TYPE_DESCRIPTION;
exports.ATTR_VTEX_OPERATION_ID = attributes.ATTR_VTEX_OPERATION_ID;
exports.ATTR_VTEX_OPERATION_ID_DESCRIPTION = attributes.ATTR_VTEX_OPERATION_ID_DESCRIPTION;
exports.ATTR_VTEX_USER_AGENT = attributes.ATTR_VTEX_USER_AGENT;
exports.ATTR_VTEX_USER_AGENT_DESCRIPTION = attributes.ATTR_VTEX_USER_AGENT_DESCRIPTION;
