// @vtex/diagnostics-nodejs/dist/samplers/traces/debug.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugSampler = void 0;
exports.newDebugSampler = newDebugSampler;
const api_1 = require("@opentelemetry/api");
const sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
const metadata_1 = require("./metadata");
const debugAttributeKeys = [
    'debug',
    'debug.enabled',
    'sampling.debug',
    'force.sampling'
];
class DebugSampler {
    shouldSample(context, traceId, spanName, spanKind, attributes, links) {
        const metadata = (0, metadata_1.newSamplerMetadata)('debug');
        for (const key of debugAttributeKeys) {
            if (attributes[key] === 'true' || attributes[key] === true) {
                metadata.decision = 'record_and_sample';
                metadata.reason = 'debug_enabled';
                metadata.extra['trigger'] = key;
                return {
                    decision: sdk_trace_base_1.SamplingDecision.RECORD_AND_SAMPLED,
                    attributes: metadata.toAttributes()
                };
            }
        }
        const parentContext = api_1.trace.getSpanContext(context);
        if (parentContext?.traceFlags && (parentContext.traceFlags & 1)) {
            metadata.decision = 'record_and_sample';
            metadata.reason = 'parent_sampled';
            return {
                decision: sdk_trace_base_1.SamplingDecision.RECORD_AND_SAMPLED,
                attributes: metadata.toAttributes()
            };
        }
        metadata.decision = 'drop';
        metadata.reason = 'debug_disabled';
        return {
            decision: sdk_trace_base_1.SamplingDecision.NOT_RECORD,
            attributes: metadata.toAttributes()
        };
    }
    toString() {
        return 'DebugSampler';
    }
}
exports.DebugSampler = DebugSampler;
function newDebugSampler() {
    return new DebugSampler();
}
