// @vtex/api/lib/service/tracing/metrics/MetricNames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BETWEEN_SCRAPES_EVENT_LOOP_LAG_PERCENTILES = exports.BETWEEN_SCRAPES_EVENT_LOOP_LAG_MAX = exports.REQUEST_RESPONSE_SIZES = exports.REQUEST_TIMINGS = exports.REQUESTS_ABORTED = exports.REQUESTS_TOTAL = exports.CONCURRENT_REQUESTS = void 0;
exports.CONCURRENT_REQUESTS = {
    name: 'io_http_requests_current',
    help: 'The current number of requests in course.',
    type: "gauge" /* METRIC_TYPES.GAUGE */,
};
exports.REQUESTS_TOTAL = {
    name: 'runtime_http_requests_total',
    help: 'The total number of HTTP requests.',
    labelNames: ["status_code" /* RequestsMetricLabels.STATUS_CODE */, "handler" /* RequestsMetricLabels.REQUEST_HANDLER */],
    type: "counter" /* METRIC_TYPES.COUNTER */,
};
exports.REQUESTS_ABORTED = {
    name: 'runtime_http_aborted_requests_total',
    help: 'The total number of HTTP requests aborted.',
    labelNames: ["handler" /* RequestsMetricLabels.REQUEST_HANDLER */],
    type: "counter" /* METRIC_TYPES.COUNTER */,
};
exports.REQUEST_TIMINGS = {
    name: 'runtime_http_requests_duration_milliseconds',
    help: 'The incoming http requests total duration.',
    labelNames: ["handler" /* RequestsMetricLabels.REQUEST_HANDLER */],
    buckets: [10, 20, 40, 80, 160, 320, 640, 1280, 2560, 5120],
    type: "histogram" /* METRIC_TYPES.HISTOGRAM */,
};
exports.REQUEST_RESPONSE_SIZES = {
    name: 'runtime_http_response_size_bytes',
    help: `The outgoing response sizes (only applicable when the response isn't a stream).`,
    labelNames: ["handler" /* RequestsMetricLabels.REQUEST_HANDLER */],
    buckets: [500, 2000, 8000, 16000, 64000, 256000, 1024000, 4096000],
    type: "histogram" /* METRIC_TYPES.HISTOGRAM */,
};
exports.BETWEEN_SCRAPES_EVENT_LOOP_LAG_MAX = {
    name: 'runtime_event_loop_lag_max_between_scrapes_seconds',
    help: 'The max event loop lag that occurred between this and the previous scrape',
    type: "gauge" /* METRIC_TYPES.GAUGE */,
};
exports.BETWEEN_SCRAPES_EVENT_LOOP_LAG_PERCENTILES = {
    name: 'runtime_event_loop_lag_percentiles_between_scrapes_seconds',
    help: 'Event loop lag percentiles from the observations that occurred between this and the previous scrape',
    labelNames: ["percentile" /* EventLoopMetricLabels.PERCENTILE */],
    type: "gauge" /* METRIC_TYPES.GAUGE */,
};
