// @vtex/api/lib/tracing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FORMAT_HTTP_HEADERS = exports.Span = exports.getTraceInfo = exports.createTracingContextFromCarrier = exports.TracingTags = exports.SpanReferenceTypes = exports.createSpanReference = exports.ErrorReport = exports.ErrorKinds = void 0;
const opentracing_1 = require("opentracing");
Object.defineProperty(exports, "FORMAT_HTTP_HEADERS", { enumerable: true, get: function () { return opentracing_1.FORMAT_HTTP_HEADERS; } });
Object.defineProperty(exports, "Span", { enumerable: true, get: function () { return opentracing_1.Span; } });
var node_error_report_1 = require("@vtex/node-error-report");
Object.defineProperty(exports, "ErrorKinds", { enumerable: true, get: function () { return node_error_report_1.ErrorKindsBase; } });
var ErrorReport_1 = require("./errorReporting/ErrorReport");
Object.defineProperty(exports, "ErrorReport", { enumerable: true, get: function () { return ErrorReport_1.ErrorReport; } });
var createSpanReference_1 = require("./spanReference/createSpanReference");
Object.defineProperty(exports, "createSpanReference", { enumerable: true, get: function () { return createSpanReference_1.createSpanReference; } });
var SpanReferenceTypes_1 = require("./spanReference/SpanReferenceTypes");
Object.defineProperty(exports, "SpanReferenceTypes", { enumerable: true, get: function () { return SpanReferenceTypes_1.SpanReferenceTypes; } });
var Tags_1 = require("./Tags");
Object.defineProperty(exports, "TracingTags", { enumerable: true, get: function () { return Tags_1.UserlandTags; } });
var UserLandTracer_1 = require("./UserLandTracer");
Object.defineProperty(exports, "createTracingContextFromCarrier", { enumerable: true, get: function () { return UserLandTracer_1.createTracingContextFromCarrier; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getTraceInfo", { enumerable: true, get: function () { return utils_1.getTraceInfo; } });
