// @vtex/diagnostics-nodejs/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewTelemetryClient = exports.Instrumentation = exports.Processors = exports.Providers = exports.Samplers = exports.Exporters = exports.Traces = exports.Metrics = exports.Logs = exports.Types = void 0;
const tslib_1 = require("tslib");
exports.Types = tslib_1.__importStar(require("./types"));
exports.Logs = tslib_1.__importStar(require("./logs"));
exports.Metrics = tslib_1.__importStar(require("./metrics"));
exports.Traces = tslib_1.__importStar(require("./traces"));
exports.Exporters = tslib_1.__importStar(require("./exporters"));
exports.Samplers = tslib_1.__importStar(require("./samplers"));
exports.Providers = tslib_1.__importStar(require("./providers"));
exports.Processors = tslib_1.__importStar(require("./processors"));
exports.Instrumentation = tslib_1.__importStar(require("./instrumentation"));
var telemetry_1 = require("./telemetry");
Object.defineProperty(exports, "NewTelemetryClient", { enumerable: true, get: function () { return telemetry_1.NewTelemetryClient; } });
