// @vtex/diagnostics-nodejs/dist/exporters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdoutExporter = exports.OTLPExporter = exports.TelemetryType = exports.CreateMetricsExporterConfig = exports.CreateLogsExporterConfig = exports.CreateTracesExporterConfig = void 0;
exports.CreateExporter = CreateExporter;
const otlp_1 = require("./otlp");
Object.defineProperty(exports, "OTLPExporter", { enumerable: true, get: function () { return otlp_1.OTLPExporter; } });
const stdout_1 = require("./stdout");
Object.defineProperty(exports, "StdoutExporter", { enumerable: true, get: function () { return stdout_1.StdoutExporter; } });
const interface_1 = require("./interface");
Object.defineProperty(exports, "CreateTracesExporterConfig", { enumerable: true, get: function () { return interface_1.CreateTracesExporterConfig; } });
Object.defineProperty(exports, "CreateLogsExporterConfig", { enumerable: true, get: function () { return interface_1.CreateLogsExporterConfig; } });
Object.defineProperty(exports, "CreateMetricsExporterConfig", { enumerable: true, get: function () { return interface_1.CreateMetricsExporterConfig; } });
function CreateExporter(config, type = 'stdout') {
    switch (type) {
        case 'otlp':
            return new otlp_1.OTLPExporter(config);
        case 'stdout':
            return new stdout_1.StdoutExporter(config);
        default:
            throw new Error(`Unsupported exporter type: ${type}`);
    }
}
var exporters_1 = require("../types/exporters");
Object.defineProperty(exports, "TelemetryType", { enumerable: true, get: function () { return exporters_1.TelemetryType; } });
